system/Database/OCI8/Forge.php

```diff
@@ -273,18 +273,22 @@ protected function _dropTable(string $table, bool $ifExists, bool $cascade)
         return $sql;
     }
 
+    /**
+     * Generates SQL to process foreign keys
+     */
     protected function _processForeignKeys(string $table): string
     {
-        $sql = '';
+        $allowActions = ['CASCADE', 'SET NULL', 'NO ACTION'];
 
-        $allowActions = [
-            'CASCADE',
-            'SET NULL',
-            'NO ACTION',
-        ];
+        $sql = '';
 
         foreach ($this->foreignKeys as $fkey) {
-            $nameIndex            = $table . '_' . implode('_', $fkey['field']) . '_fk';
+            $nameIndex = $table . '_' . implode('_', $fkey['field']) . '_foreign';
+
+            if (version_compare($this->db->getVersion(), '12.2.0', '<') && strlen($nameIndex) > 30) {
+                $nameIndex = substr($nameIndex, 0, 28) . mt_rand(10, 99);
+            }
```

Comment:
Do you really need this?
I guess it is better not to do like this.
If you have a long table name, and mt_rand() returns the same value,
$nameIndex would be the same name.

Reply:
>Personally I hope an error occurs if the length is more than the database limit.

This isn't helpful because then Forge could never create the index.

mt_rand() is just an extra insurance policy. Its possible to return same number but highly unlikely that this happens and there is also a collision of names.

Ideally we add the ability to specify the names manually. I know there are some issues related to this. I planned to take a look at doing this in the future. Hopefully before the release of 4.3 we can add this and then get rid of this piece of code. But for now this should keep things working.

Reply:
We cannot create an index now with Forge if the index name length exceeds the database limit.
So it is not a problem.  It is the same as the current behavior.

This check is a breaking change. And if we remove this check, it will be another breaking change.
4.3 will  be released soon.

I don't like naming with random number. It is very bad naming. Humans can't get it.
If you plan to get rid of the code later, I hope you don't add it.

If you have the need of the code, please customize your own Forge.
I do not see why this code must be added.

Reply:
>I do not see why this code must be added.

Because under certain circumstances a user of Oracle won't be able to create indexes without it. It would likely rarely get invoked but in that special case it wouldn't be a road block. Even for older versions the code isn't invoked unless its certain to fail. You can remove it.. then the developer will just have to manually create the index without using Forge.

Reply:
```php
            if (version_compare($this->db->getVersion(), '12.2.0', '<') && strlen($nameIndex) > 30) {
                $nameIndex = uniqid();
            }
```

Reply:
>This check is a breaking change. And if we remove this check, it will be another breaking change.

This was broken before anything was done here.

Reply:
It doesn't appear to me that it has ever been documented what the names of keys will be, Honestly If I were working with keys I would look them up and identify them by the tables and columns used.

Reply:
> This was broken before anything was done here.

What do you mean? Before this PR, too long index name causes an error, but this PR does not cause an error and silently changes the naming convention. I think this is a breaking change.
If my understanding is incorrect, please point it out.

Reply:
Devs can create any indexes with `$db->query()`.
So it is not crucial that Forge cannot create indexes in some cases.

Reply:
Ok, if you would rather it create an error then I'll remove the code. I feel that the index name is rather inconsequential to the developer. It does help out the framework though to have consistent names for testing. I feel that having forge working and not throwing an error is more important then what name is given. The documentation makes no express or implied warranties as to the name that will be used but it is implied that it will create a foreign key. By allowing the error you are breeching the implied warranty.

This was broken before because a naming scheme was chosen and put into use that could inevitably cause errors in certain circumstances. This is true not for just OCI8 but the rest as well. One of the challenges of having the framework name the constraints is ensuring unique names are generated. A method was chosen that insured this outcome but unfortunately it isn't compatible all the time. What my code does is try to stick to the naming scheme but ultimately insure a unique name is applied. 

I would have changed them all to a suffix of _fk instead of _foreign but this would have created more changes then the alternative. Oracle is probably the least used database of them all and foreign keys are probably the least used indexes so hopefully the change doesn't cause too many issues.

SQLite doesn't even assign a name to the foreign keys. Probably because you can't do anything with them other than drop and create a new table. This PR however assigns them a name based on the naming scheme which allows us to handle them as if they are named. This makes it easier to drop and add them.

Let me know if you just want to drop this bit of code and allow the error to occur.

Reply:
The random name is not consistent.
In any cases, naming is very important. It is a core of abstraction.
Bad naming should be always avoided.

I oppose the code in this PR. It is too adhoc. I cannot accept it. Please remove it.
If you want to try to include such code to avoid errors into CI4, send another PR only for it.
It is a big issue. We need more discussion.

Reply:
>In any cases, naming is very important. It is a core of abstraction.

Well this PR tries to give them a consistent name. And in the event it cannot provide exactly it comes as close to it at possible. 

Do the math. You have 30 characters to create a unique name. the sum of all the tables, columns and database prefix is more than 30 characters. 

And if the developer is allowed to set their own names then that doesn't further your cause.

Reply:
You have a point. It seems changing to `_foreign` is not good for Oracle 12.1.
It is better to keep `_fk` for  Oracle 12.1.

Reply:
Well to be consistent Id change them all to _fk. But that still doesn't solve the issue where the name is too long. In some cases a different scheme is required one way or another. 

Its a well documented issue. Oracle makes having a table column based naming scheme impossible sometimes.

https://stackoverflow.com/questions/199498/foreign-key-naming-scheme

Reply:
Looking through my databases it looks like my longest table name is 37 characters.